packages/gatsby/src/utils/parcel/compile-gatsby-files.ts

```diff
@@ -61,24 +61,36 @@ export async function compileGatsbyFiles(
   retry: number = 0
 ): Promise<void> {
   try {
+    const configName = `gatsby-node`
+
     // Check for gatsby-node.jsx and gatsby-node.tsx (or other misnamed variations)
-    const files = await readdir(siteRoot)
+    const filesAndDirectories = await readdir(siteRoot, { withFileTypes: true })
+    const files = filesAndDirectories
+      .filter(dirent => !dirent.isDirectory())
+      .map(dirent => dirent.name)
+
+    // Check if they have a gatsby-node directory for lifecycle function code.
+    const hasGatsbyNodeDir =
+      filesAndDirectories.filter(
+        dirent => dirent.isDirectory() && dirent.name === configName
+      ).length > 0
```

Comment:
Isn't this doing a bunch of duplicate work? This can potentially iterate twice (with the `filter`) over a lot of files, so let's combine this into just one run?

Reply:
Yea, maybe.  How would we combine these?  First is filtering out directories, second is for determining if there is a directory called `gatsby-node/`.